### PR TITLE
Expose dynamic symbols in ELF files

### PR DESCRIPTION
**Pull request by gabrielesvelto**
This fixes issue #420

### CODE REVIEW

**Comment by gabrielesvelto** [comment]
This is a crude fix to expose the dynamic symbols. It provides improved results when dumping Firefox' libxul.so as well as when dumping symbols for ELF libraries that have been stripped and for which external debug information is not avaialble (such as those provided in Arch packages or on Android devices).

**Review by Swatinem — APPROVED** [review]

File: symbolic-debuginfo/src/elf.rs [hunk 1]
```diff
@@ -214,14 +217,16 @@ impl<'data> ElfObject<'data> {
 
     /// Determines whether this object exposes a public symbol table.
     pub fn has_symbols(&self) -> bool {
-        !self.elf.syms.is_empty()
+        !self.elf.syms.is_empty() && !self.elf.dynsyms.is_empty()
```
Comment (Swatinem) [inline symbolic-debuginfo/src/elf.rs:220]:
```suggestion
        !self.elf.syms.is_empty() || !self.elf.dynsyms.is_empty()
```

Easy to miss, but this failed the testsuite:
```
   12       │-        has_symbols: true,
         12 │+        has_symbols: false,
```

Comment (loewenheim) [inline symbolic-debuginfo/src/elf.rs:220]: Why do both `syms` and `dynsyms` have to be nonempty here?

File: symbolic-debuginfo/src/elf.rs [hunk 2]
```diff
@@ -550,37 +557,60 @@ impl<'data, 'object> Iterator for ElfSymbolIterator<'data, 'object> {
...
+        ) {
+            Some(debug_symbol)
+        } else {
+            get_symbols(
+                &mut self.dynamic_symbols,
+                self.dynamic_strtab,
+                self.load_addr,
+                self.sections,
+            )
+        }
```
Comment (loewenheim) [inline symbolic-debuginfo/src/elf.rs:614]: Personally, I think this reads better with `or_else`:
```suggestion
        get_symbols(
            &mut self.symbols,
            self.strtab, 
            self.load_addr,
            self.sections,
        )
        .or_else(|| {
            get_symbols(
                &mut self.dynamic_symbols,
                self.dynamic_strtab,
                self.load_addr,
                self.sections,
            )
        })
```

File: symbolic-debuginfo/src/elf.rs (hunk 1 above)
Comment (gabrielesvelto) [reply symbolic-debuginfo/src/elf.rs:220]: Ah yes, silly mistake on my part

File: symbolic-debuginfo/src/elf.rs (hunk 2 above)
Comment (gabrielesvelto) [reply symbolic-debuginfo/src/elf.rs:614]: Yes, much better. I'll put this change into the patch and push again. I also considered chaining `self.elf.syms` and `self.elf.dynsyms` when iterating over the symbols but then I couldn't find a simple way to select between `strtab` and `dynstrtab` so I went for this cruder approach.

**Review by loewenheim — APPROVED** [review]

**Review by Swatinem — APPROVED** [review]

**Pull request merged**